### PR TITLE
Close files when CachingFileManager is garbage collected

### LINKED ISSUE #2560: Lock related problem in on travis-ci but not on local machine
There is a `KeyError` at `xarray.backends.file_manager.CachingFileManager` for the `acquire` method on these lines:

```
        with self._lock:
            try:
                file = self._cache[self._key]
```

which does not happen when I test on my macos.

Let me know if you require further testing on my part.

Full log:

```
=================================== FAILURES ===================================

________________________________ test_qtl_xarr _________________________________


self = <xarray.backends.file_manager.CachingFileManager object at 0x7f59e2cab978>


    def acquire(self):

        """Acquiring a file object from the manager.

    

        A new file is only opened if it has expired from the

        least-recently-used cache.

    

        This method uses a reentrant lock, which ensures that it is

        thread-safe. You can safely acquire a file in multiple threads at the

        same time, as long as the underlying file object is thread-safe.

    

        Returns

        -------

        An open file object, as returned by ``opener(*args, **kwargs)``.

        """

        with self._lock:

            try:

>               file = self._cache[self._key]


miniconda/envs/test-environment/lib/python3.6/site-packages/xarray/backends/file_manager.py:137: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 


self = <xarray.backends.lru_cache.LRUCache object at 0x7f5a04a9c630>

key = [<function _open_netcdf4_group at 0x7f5a04abc1e0>, ('/tmp/tmprj8lwlat/xarr.hdf5', CombinedLock([<SerializableLock: 447...>])), 'r', (('clobber', True), ('diskless', False), ('format', 'NETCDF4'), ('group', '/foo/chr1'), ('persist', False))]


    def __getitem__(self, key):

        # record recent use of the key by moving it to the front of the list

        with self._lock:

>           value = self._cache[key]

E           KeyError: [<function _open_netcdf4_group at 0x7f5a04abc1e0>, ('/tmp/tmprj8lwlat/xarr.hdf5', CombinedLock([<SerializableLock: 4471090f-419e-49bd-9d27-4156e09febdd>, <SerializableLock: b9eb5ba1-2830-402e-b48f-3ef45768b7c2>])), 'r', (('clobber', True), ('diskless', False), ('format', 'NETCDF4'), ('group', '/foo/chr1'), ('persist', False))]


miniconda/envs/test-environment/lib/python3.6/site-packages/xarray/backends/lru_cache.py:43: KeyError


During handling of the above exception, another exception occurred:


    def test_qtl_xarr():

        with limix.example.file_example("xarr.hdf5.bz2") as filepath:

            filepath = limix.sh.extract(filepath, verbose=False)

            sample_ids = limix.io.hdf5.fetch(filepath, "/foo/chr1/col_header/sample_ids")

    

            rsid = dict()

            for i in range(1, 3):

                rsid[i] = limix.io.hdf5.fetch(

                    filepath, "/foo/chr{}/row_header/rsid".format(i)

                )

    

            G = []

            for i in range(1, 3):

>               g = xr.open_dataset(filepath, "/foo/chr{}".format(i))["matrix"]


miniconda/envs/test-environment/lib/python3.6/site-packages/limix/qtl/test/test_qtl_xarr.py:20: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

miniconda/envs/test-environment/lib/python3.6/site-packages/xarray/backends/api.py:320: in open_dataset

    filename_or_obj, group=group, lock=lock, **backend_kwargs)

miniconda/envs/test-environment/lib/python3.6/site-packages/xarray/backends/netCDF4_.py:355: in open

    return cls(manager, lock=lock, autoclose=autoclose)

miniconda/envs/test-environment/lib/python3.6/site-packages/xarray/backends/netCDF4_.py:314: in __init__

    self.format = self.ds.data_model

miniconda/envs/test-environment/lib/python3.6/site-packages/xarray/backends/netCDF4_.py:359: in ds

    return self._manager.acquire().value

miniconda/envs/test-environment/lib/python3.6/site-packages/xarray/backends/file_manager.py:143: in acquire

    file = self._opener(*self._args, **kwargs)

miniconda/envs/test-environment/lib/python3.6/site-packages/xarray/backends/netCDF4_.py:247: in _open_netcdf4_group

    ds = nc4.Dataset(filename, mode=mode, **kwargs)

netCDF4/_netCDF4.pyx:2135: in netCDF4._netCDF4.Dataset.__init__

    ???

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 


>   ???

E   OSError: [Errno -101] NetCDF: HDF error: b'/tmp/tmprj8lwlat/xarr.hdf5'


netCDF4/_netCDF4.pyx:1752: OSError

===================== 1 failed, 50 passed in 65.82 seconds =====================

The command "bash <(curl -fsSL https://raw.githubusercontent.com/horta/ci/master/travis.sh)" exited with 1.




Done. Your build exited with 1.
```

### LINKED ISSUE #2614: Depending upon how compute() is called, python seg faults
I have two large dask arrays (`hist_data_ua850` and `hist_data_hus850`). If I want to calculate the product then `(hist_data_ua850 * hist_data_hus850).compute()` fails with either seg fault or bus error. However, `(hist_data_ua850.compute() * hist_data_hus850.compute())` works. (sample code attached
)

The `gdb` produce following error message for seg fault.
```
Thread 25 "python3" received signal SIGBUS, Bus error.
[Switching to Thread 0x7fffbf5e3700 (LWP 31903)]
0x00007fffec92a8de in H5FL_fac_malloc ()
   from /home/user/.local/lib/python3.6/site-packages/netCDF4/.libs/libhdf5-d5cdc8a0.so.101.1.0                          
```
or 
```
Thread 25 "python3" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fffbf5e3700 (LWP 11553)]
0x00007fffeca3cb57 in ?? () from /home/user/.local/lib/python3.6/site-packages/netCDF4/.libs/libhdf5-d5cdc8a0.so.101.1.0
```

During these seg fault <10% of RAM is being used.
#### Output of ``xr.show_versions()``

<details>
INSTALLED VERSIONS                        
------------------                          
commit: None                              
python: 3.6.8.candidate.1                       
python-bits: 64                                                 
OS: Linux                                                          
OS-release: 4.18.0-3-amd64                                                                                                
machine: x86_64                           
processor:                                
byteorder: little                              
LC_ALL: en_IN.UTF-8                       
LANG: en_IN.UTF-8                         
LOCALE: en_IN.UTF-8                               
                                          
xarray: 0.11.0                            
pandas: 0.23.4                            
numpy: 1.15.4                                                                         
scipy: 1.1.0                              
netCDF4: 1.4.2      
h5netcdf: None
h5py: 2.8.0
Nio: None
zarr: None
cftime: 1.0.3.4
PseudonetCDF: None
rasterio: 1.0.2
iris: None
bottleneck: 1.2.1
cyordereddict: None
dask: 1.0.0
distributed: 1.20.2
matplotlib: 3.0.2
cartopy: 0.17.0
seaborn: 0.9.0
setuptools: 40.2.0
pip: 18.0
conda: None
pytest: 3.4.0
IPython: 7.2.0
sphinx: 1.8.2

</details>

### PR DESCRIPTION
**Pull request by shoyer**
This frees users from needing to worry about this.

Using `__del__` turned up to be easier than  using weak references.

 - [x] Closes #2560 
 - [x] Closes #2614
 - [x] Tests added
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new AP

### CODE REVIEW

**Comment by shoyer** [comment]
Of course everything runs fine either on my laptop (Macbook) or work machine (Linux). I wonder what's going on with Travis-CI here.

**Comment by shoyer** [comment]
@pydata/xarray It would be great if someone else could take a look at this, though certainly thread-safety and garbage collection are slightly painful concepts to think deeply about. If nobody reviews this, I will probably self-merge in a few days.

File: xarray/backends/locks.py [hunk 1]
```diff
@@ -113,6 +113,27 @@ def get_write_lock(key):
...
+    elif DistributedLock is not None and isinstance(lock, DistributedLock):
+        # distributed.Lock doesn't support the blocking argument yet:
+        # https://github.com/dask/distributed/pull/2412
+        return lock.acquire(timeout=0)
+    else:
+        # "blocking" keyword argument not supported for:
+        # - threading.Lock on Python 2.
+        # - dask.SerializableLock with dask v1.0.0 or earlier.
+        # - multiprocessing.Lock calls the argument "block" instead.
+        return lock.acquire(blocking)
```
Comment (jhamman) [inline xarray/backends/locks.py]: Given the date, do you think its useful here to consider this Python 2 work around? I guess its the same work-around as the dask/multiprocessing issue.

**Comment by shoyer** [comment]
Yeah, it's not really any additional trouble. I plan to include this in a
v0.11.1 release, which will be the last release to support Python 2.

**Comment by shoyer** [comment]
OK, I'm actually going to merge soon (and issue the 0.11.1 release) unless I hear any objections...

**Pull request merged**